### PR TITLE
Add verify method now we can verify single request

### PR DESCRIPTION
**Pull request by rocketb**
With this feature, we can verify single request without mocking. It will be helpful in some proxy cases.

Something like that:

```python
from mockserver import MockClient, request, times

client = MockClient('localhost:1080')
client.verify_request(request(method='GET', path='/path', querystring={'test': 'test'}), timing=times(1))
```

### CODE REVIEW

File: mockserver/__init__.py [hunk 1]
```diff
@@ -27,13 +27,16 @@ def expect(self, request, response, timing, time_to_live=None):
         self.stub(request, response, timing, time_to_live)
         self.expectations.append((request, timing))
 
-    def verify(self):
+    def verify(self, request, timing=1):
```
Comment (lindycoder) [inline mockserver/__init__.py]: timing shall now have a default, like the expect it think it's important to be specified, plus you put an INT there, the int won't have a for_verification method :P

File: test/test_basic_expectations.py [hunk 2]
```diff
@@ -12,7 +12,7 @@ def test_expect_once_not_called_fails(self):
         )
 
         with self.assertRaises(AssertionError):
-            self.client.verify()
+            self.client.verify_expectations()
```
Comment (lindycoder) [inline test/test_basic_expectations.py:15]: could you change these to teardown as they are testing the `expect` behavior

**Review by lindycoder — CHANGES_REQUESTED** [review]
All right, i can see the use case.  This will be a contract change so let's make it clear.

the `verify` method is then a new method of this client's API, it should have its own test suite

I propose `test_basic_verifying`.

There are 2 ways i would do this:

1 - Bit more complex : since the goal of the tests here are to be the documentation maybe it would make sense to use mock server as a proxy to make the test.  To do this you would probably need a second service in the docker-compose called proxy_server with special command that adds a -Dhttp.proxyHost=mock-server that proxies to the other container.   Then in the setup of the test i would stub a call on the mock-server and in the test: requests.get on the proxy-server and do a .verify on the proxy-server.
That would show how to use the proxy and document the use case and show the .verify usage.

2 - Much simplier
You can make a simple test that .stub and .verify, though i would put the .stub in the setup so that it is clearly not part of the intent of the test.

So there, you can choose whichever path you want :)

**Comment by rocketb** [comment]
So with your agreement, I chose path 3 :sweat_smile:. In case that we don't check stubing and verifying works for hits of MockServer, we can just hit the mock server with any request, so we don't need .stub in setUp step. See [Mock server Verifications](http://mock-server.com/mock_server/verification.html)

I move some expectations checks in verify test class, because now .verify_expectations just iterate throw all items in expectations dict, please check it out.

File: test/test_basic_verifying.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import requests
+from mockserver import request, response, times
+from test import MOCK_SERVER_URL, MockServerClientTestCase
+
+
+class TestBasicVerifying(MockServerClientTestCase):
+    def test_verify_request_recived_once(self):
```
Comment (lindycoder) [inline test/test_basic_verifying.py]: you got a typo, it's `received`, the next 2 tests also have the typo

File: test/test_basic_expectations.py [hunk 4]
```diff
@@ -4,16 +4,6 @@
 
 
 class TestBasicExpectations(MockServerClientTestCase):
-    def test_expect_once_not_called_fails(self):
```
Comment (lindycoder) [inline test/test_basic_expectations.py]: i understand that it is tested in the verify but this serve as the documentation and i like having a clear test showing : 
if you expect 1 and you verify_expectations you will get that assertion error, same goes for the other test.

The intent is really to show the usage

**Review by lindycoder — CHANGES_REQUESTED** [review]
Oh you're right you can verifying calls that aren't stubbed at all, my bad!

But since this is a contract change there's a change to be done in the readme in the expect section it should invoke .verify_expectations

you can add a section for verify alone if you feel like it

**Comment by rocketb** [comment]
Ok, you're right.
That's one little update, one step closer to the goal. )

**Comment by lindycoder** [comment]
There's a wierd flakyness on this test

FAIL: test_verify_all_expectations (test.test_basic_verifying.TestBasicVerifying)

It failed the current check on 2.7 and 3.4 and before your last change it also failed on 2.7

does it pass locally for you?

**Comment by lindycoder** [comment]
it's as if the mock-server was not cleaned properly between 2 tests... as there's a stub on /path that lingers?  

everything is fine when i run that locally...  so wierd

**Comment by rocketb** [comment]
Yep, locally it is fine... 
I confirm somehow mock server doesn't clean up on CI

**Comment by lindycoder** [comment]
all right i made another PR based on yours and did some experimentation and here's my result:

add     
```
def tearDown(self):
        self.client.reset()
```
in test/__init__.py, it _seems_ to resolve most test pollution issues.

After that there's only this test that is very very flaky:
`test_verify_request_received_once (test.test_basic_verifying.TestBasicVerifying)`
saying
`AssertionError: Request not found exactly once, expected:<{ }> but was:<>`

as if the request never happened, so it looks a lot like race condition

Not sure how to address this:
- `time.sleep(0.1)` in the test looks very bad...
- we could add a `@retry(times=5, delay 0.02)` on the `client.verify` method but that forces the verify to have unexpected behaviors forced on the users.
- we could `retry_call(self.client.verify, args=(request(), times(1)), times=5, delay 0.1)` directly in the test

Do you have an opinion on this matter?

**Comment by Marx314** [comment]
or within the setup of MockServerClientTestCase but after the reset

add a call to retrieve?type=RECORDED_EXPECTATIONS until it's empty?

-> http://www.mock-server.com/proxy/record_and_replay.html

**Comment by lindycoder** [comment]
@Marx314 Nice to see you!

doing that could solve the problem of having to put a reset in the teardown, but that won't help the tests that does `request.get` directly followed by `verify`that whines it doesn't have received a request yet.

**Comment by rocketb** [comment]
Since this is network or MockServer issue but not the bug of client lib, the best way on my opinion will be using mocks in future.
For now, I suggest adding call of retrieve in setUp and the pooling of .verify method.

File: mockserver/__init__.py [hunk 5]
```diff
@@ -8,8 +8,8 @@ def __init__(self, base_url):
         self.base_url = base_url
         self.expectations = []
 
-    def _call(self, command, data=None):
-        return requests.put("{}/{}".format(self.base_url, command), data=data)
+    def _call(self, command, data=None, **kwargs):
```
Comment (lindycoder) [inline mockserver/__init__.py:11]: i personnally don't enjoy kwargs in code we control, i think it obfuscates usage, it's a pain to debug what's coming in and out, and screws auto completion / parameters inferences

**Review by lindycoder — CHANGES_REQUESTED** [review]
I get what you try to do with the retrieve but that only asserts that the reset works.

The most problematic test we seem to have is the one where the request.get is somewhat slower than the .verify and the retreive does not fix that.

The only use i would see for the retrieve would be to show what's left on test failure which would need to be in the tearDown and i remember trying to determine whether a test was failing int he tearddown and having troubles with it.

So if you really want the retrieval i'd put this is a separate PR.

As for mocks vs the real thing, i was hoping testing with the container provides a fast enough feedback loop to act as unit tests and it tests quirks that could happen in the api, like that time i had to put in an Ordered Dict because the params were not read correctly.  Bonus of this method : we'll see versions contract breaks if there are any.

Now for the next step i'll try to get the PR again and work with it a bit...

**Comment by lindycoder** [comment]
All right let's try stuff,

could you cherry pick or rebase onto #11  that might give us some more insight on the problem.

(And, somewhat annoying, i haven't been able to reproduce the flakyness in --verbose mode... :P)

**Comment by lindycoder** [comment]
interesting... on that last failure:

the test does
- reset
- expect
- request
- verify

but the logs from the container sees:
- reset
- expect
- request
- RESET
- verify

and that extra reset is the setup of the _next_ test.
So somehow the next test starts before that one finished!?

**Comment by lindycoder** [comment]
All right i got it working!

Check #15 that will stabilize tests a lot.

I created #14 with your first 3 commits and a little addition, stubbing the first verify test and i wasn't able to make it fail yet.

So you can either rebase off #15, squash your first 3 commit along with the stubbing i added in #14

Or we can merge #14, what do you think?


If you want the retrieve functionality we can make a new PR but i'd rethink the api a bit.

Thanks for your contribution, sorry that i made it complicated to make reliable tests, i think testing the container is worth it :)

**Comment by lindycoder** [comment]
Have been replaced by #14

**Pull request closed without merging**